packages/backend-core/src/objectStore/objectStore.ts

```diff
@@ -388,50 +415,61 @@ export async function getPresignedUrl(
-  if (data instanceof stream.Readable) {
-    data.pipe(fs.createWriteStream(outputPath))
-  } else {
-    fs.writeFileSync(outputPath, data)
-  }
-  return outputPath
+  return await tracer.trace("retrieveToTmp", async span => {
+    span.addTags({ bucketName, filepath })
+    bucketName = sanitizeBucket(bucketName)
+    filepath = sanitizeKey(filepath)
+    const data = await retrieve(bucketName, filepath)
+    const outputPath = join(budibaseTempDir(), v4())
+    span.addTags({ outputPath })
+    if (data instanceof stream.Readable) {
+      span.addTags({ stream: true })
+      data.pipe(fs.createWriteStream(outputPath))
+    } else {
+      span.addTags({ stream: false })
+      fs.writeFileSync(outputPath, data)
+    }
+    return outputPath
+  })
 }
 
 export async function retrieveDirectory(bucketName: string, path: string) {
-  let writePath = join(budibaseTempDir(), v4())
-  fs.mkdirSync(writePath)
-  const objects = await listAllObjects(bucketName, path)
-  let streams = await Promise.all(
-    objects.map(obj => getReadStream(bucketName, obj.Key!))
-  )
-  let count = 0
-  const writePromises: Promise<void>[] = []
-  for (let obj of objects) {
-    const filename = obj.Key!
-    const stream = streams[count++]
-    const possiblePath = filename.split("/")
-    const dirs = possiblePath.slice(0, possiblePath.length - 1)
-    const possibleDir = join(writePath, ...dirs)
-    if (possiblePath.length > 1 && !fs.existsSync(possibleDir)) {
-      fs.mkdirSync(possibleDir, { recursive: true })
-    }
-    const writeStream = fs.createWriteStream(join(writePath, ...possiblePath), {
-      mode: 0o644,
-    })
-    stream.pipe(writeStream)
-    writePromises.push(
-      new Promise<void>((resolve, reject) => {
+  return await tracer.trace("retrieveDirectory", async span => {
+    span.addTags({ bucketName, path })
+
+    let writePath = join(budibaseTempDir(), v4())
+    await fsp.mkdir(writePath)
+
+    let numObjects = 0
+    for await (const object of listAllObjects(bucketName, path)) {
+      numObjects++
+
+      const filename = object.Key!
+      const stream = await getReadStream(bucketName, filename)
+      const possiblePath = filename.split("/")
+      const dirs = possiblePath.slice(0, possiblePath.length - 1)
+      const possibleDir = join(writePath, ...dirs)
+      if (possiblePath.length > 1 && !fs.existsSync(possibleDir)) {
+        await fsp.mkdir(possibleDir, { recursive: true })
+      }
+      const writeStream = fs.createWriteStream(
+        join(writePath, ...possiblePath),
+        {
+          mode: 0o644,
+        }
+      )
+
+      stream.pipe(writeStream)
+      await new Promise<void>((resolve, reject) => {
         writeStream.on("finish", () => resolve())
         stream.on("error", reject)
         writeStream.on("error", reject)
       })
-    )
-  }
-  await Promise.all(writePromises)
-  return writePath
+    }
+
+    span.addTags({ numObjects })
+    return writePath
+  })
```

Comment:
This function has been reworked. `listAllObjects` is now an async generator, to prevent loading thousands of file names into memory at once, and instead of gathering all promises together and calling Promise.all on them, we await them individually.